# Análise exploratória de dados

In [1]:
import pandas as pd
import os
from termcolor import colored
from IPython.display import display
import numpy as np
import seaborn as sns
import matplotlib.pyplot as plt

In [2]:

caminho = '../data/raw'

# Leitura dos conjuntos de dados
tabelas = {
    os.path.splitext(arquivo)[0]: pd.read_csv(os.path.join(caminho, arquivo))
    for arquivo in os.listdir(caminho)
    if arquivo.endswith('.csv')
}

# Exibir
for nome, df in tabelas.items():
    print(colored(f'\n=== {nome.upper()} ===', 'blue', attrs=['bold']))
    print(f'Shape: {df.shape}')
    print('Nulos por coluna:')
    print(df.isnull().sum())
    display(df.head(3))



=== DELIVERIES ===
Shape: (378843, 5)
Nulos por coluna:
delivery_id                     0
delivery_order_id               0
driver_id                   15886
delivery_distance_meters       73
delivery_status                 0
dtype: int64


delivery_id  delivery_order_id  driver_id  delivery_distance_meters  \
0      2174658           68413340     8378.0                    5199.0   
1      2174660           68414309     2473.0                     410.0   
2      2174661           68416230     7615.0                    3784.0   

  delivery_status  
0       DELIVERED  
1       DELIVERED  
2       DELIVERED


=== STORES ===
Shape: (951, 7)
Nulos por coluna:
store_id              0
hub_id                0
store_name            0
store_segment         0
store_plan_price    115
store_latitude       16
store_longitude      16
dtype: int64


store_id  hub_id         store_name store_segment  store_plan_price  \
0         3       2            CUMIURI          FOOD               0.0   
1         6       3  PIMGUCIS DA VIVA           FOOD               0.0   
2         8       3          RASMUR S           FOOD               0.0   

   store_latitude  store_longitude  
0             NaN              NaN  
1      -30.037415        -51.20352  
2      -30.037415        -51.20352


=== ORDERS ===
Shape: (368999, 29)
Nulos por coluna:
order_id                                  0
store_id                                  0
channel_id                                0
payment_order_id                          0
delivery_order_id                         0
order_status                              0
order_amount                              0
order_delivery_fee                        0
order_delivery_cost                    7205
order_created_hour                        0
order_created_minute                      0
order_created_day                         0
order_created_month                       0
order_created_year                        0
order_moment_created                      0
order_moment_accepted                  9461
order_moment_ready                    25106
order_moment_collected                42894
order_moment_in_expedition            67429
order_moment_delivering               25316
order_moment_delivered               349398
order_moment_finished 

order_id  store_id  channel_id  payment_order_id  delivery_order_id  \
0  68405119      3512           5          68405119           68405119   
1  68405123      3512           5          68405123           68405123   
2  68405206      3512           5          68405206           68405206   

  order_status  order_amount  order_delivery_fee  order_delivery_cost  \
0     CANCELED          62.7                 0.0                  NaN   
1     CANCELED          62.7                 0.0                  NaN   
2     CANCELED         115.5                 0.0                  NaN   

   order_created_hour  ...  order_moment_delivering  order_moment_delivered  \
0                   0  ...                      NaN                     NaN   
1                   0  ...                      NaN                     NaN   
2                   0  ...                      NaN                     NaN   

   order_moment_finished  order_metric_collected_time  \
0                    NaN                          NaN   
1                    NaN                          NaN   
2                    NaN                          NaN   

  order_metric_paused_time order_metric_production_time  \
0                      NaN                          NaN   
1                      NaN                          NaN   
2                      NaN                          NaN   

  order_metric_walking_time order_metric_expediton_speed_time  \
0                       NaN                               NaN   
1                       NaN                               NaN   
2                       NaN                               NaN   

  order_metric_transit_time order_metric_cycle_time  
0                       NaN                     NaN  
1                       NaN                     NaN  
2                       NaN                     NaN  

[3 rows x 29 columns]


=== DRIVERS ===
Shape: (4824, 3)
Nulos por coluna:
driver_id       0
driver_modal    0
driver_type     0
dtype: int64


driver_id driver_modal        driver_type
0        133      MOTOBOY  LOGISTIC OPERATOR
1        138      MOTOBOY          FREELANCE
2        140      MOTOBOY          FREELANCE


=== HUBS ===
Shape: (32, 6)
Nulos por coluna:
hub_id           0
hub_name         0
hub_city         0
hub_state        0
hub_latitude     0
hub_longitude    0
dtype: int64


hub_id        hub_name      hub_city hub_state  hub_latitude  hub_longitude
0       2   BLUE SHOPPING  PORTO ALEGRE        RS    -30.047415     -51.213510
1       3  GREEN SHOPPING  PORTO ALEGRE        RS    -30.037415     -51.203520
2       4    RED SHOPPING  PORTO ALEGRE        RS    -30.021948     -51.208382


=== PAYMENTS ===
Shape: (400834, 6)
Nulos por coluna:
payment_id            0
payment_order_id      0
payment_amount        0
payment_fee         175
payment_method        0
payment_status        0
dtype: int64


payment_id  payment_order_id  payment_amount  payment_fee payment_method  \
0     4427917          68410055          118.44         0.00        VOUCHER   
1     4427918          68410055          394.81         7.90         ONLINE   
2     4427941          68412721          206.95         5.59         ONLINE   

  payment_status  
0           PAID  
1           PAID  
2           PAID


=== CHANNELS ===
Shape: (40, 3)
Nulos por coluna:
channel_id      0
channel_name    0
channel_type    0
dtype: int64


channel_id channel_name channel_type
0           1  OTHER PLACE  OWN CHANNEL
1           2  PHONE PLACE  OWN CHANNEL
2           3  WHATS PLACE  OWN CHANNEL

# 📘 Glossário dos Conjuntos de Dados

---

## 🚚 **deliveries.csv**

| Coluna                    | Descrição                                                           |
|---------------------------|----------------------------------------------------------------------|
| `delivery_id`             | Identificador único da entrega.                                     |
| `delivery_order_id`       | ID da ordem associada à entrega.                                     |
| `driver_id`               | ID do motorista responsável pela entrega.                           |
| `delivery_distance_meters` | Distância percorrida na entrega (em metros).                        |
| `delivery_status`         | Status da entrega (ex: `DELIVERED`, `CANCELLED`).                   |

---

## 🏬 **stores.csv**

| Coluna            | Descrição                                                           |
|--------------------|----------------------------------------------------------------------|
| `store_id`         | Identificador único da loja.                                         |
| `hub_id`           | ID do hub logístico ao qual a loja está associada.                  |
| `store_name`       | Nome da loja.                                                       |
| `store_segment`    | Segmento de mercado da loja (ex: FOOD, RETAIL).                     |
| `store_plan_price` | Valor do plano da loja na plataforma.                               |
| `store_latitude`   | Latitude da localização da loja.                                     |
| `store_longitude`  | Longitude da localização da loja.                                    |

---

## 📦 **orders.csv**

| Coluna               | Descrição                                                           |
|------------------------|----------------------------------------------------------------------|
| `order_id`             | Identificador único do pedido.                                     |
| `store_id`             | ID da loja onde o pedido foi feito.                                |
| `channel_id`           | Canal pelo qual o pedido foi realizado.                            |
| `payment_order_id`     | ID da transação de pagamento associada.                            |
| `delivery_order_id`    | ID da entrega associada ao pedido.                                 |

---

## 🌐 **channels.csv**

| Coluna           | Descrição                                                           |
|--------------------|----------------------------------------------------------------------|
| `channel_id`       | Identificador único do canal de vendas.                             |
| `channel_name`     | Nome do canal (ex: aplicativo, site, delivery parceiro).            |

---

## 👨‍✈️ **drivers.csv**

| Coluna           | Descrição                                                           |
|--------------------|----------------------------------------------------------------------|
| `driver_id`        | Identificador único do motorista.                                   |
| `driver_name`      | Nome do motorista.                                                  |
| `driver_vehicle`   | Tipo de veículo utilizado (ex: moto, carro, bike).                 |

---

## 🏢 **hubs.csv**

| Coluna         | Descrição                                                           |
|------------------|----------------------------------------------------------------------|
| `hub_id`         | Identificador único do centro de distribuição.                      |
| `hub_name`       | Nome do hub logístico.                                              |
| `hub_region`     | Região ou cidade do hub.                                            |

---

## 💳 **payments.csv**

| Coluna               | Descrição                                                           |
|------------------------|----------------------------------------------------------------------|
| `payment_order_id`     | Identificador único da transação de pagamento.                     |
| `payment_method`       | Método de pagamento (ex: cartão, PIX, dinheiro).                   |
| `payment_status`       | Status do pagamento (ex: `PAID`, `FAILED`).                        |
| `payment_value`        | Valor total da transação.                                          |

---

## 1. Descrição dos dados

<p align="center">
    
![Imagem do projeto](https://raw.githubusercontent.com/nosbielcs/verao_usp_2017/master/exploratoria/data_model.png)

O conjunto de dados pode ser acessado [neste link do Kaggle](https://www.kaggle.com/datasets/nosbielcs/brazilian-delivery-center). Este conjunto possui **7 datasets**, organizados da seguinte forma:

- **`channels`**: informações sobre os canais de venda (marketplaces) onde são vendidos os *good* e *food* dos lojistas.
- **`deliveries`**: informações sobre as entregas realizadas pelos entregadores.
- **`drivers`**: informações sobre os entregadores.
- **`hubs`**: informações sobre os hubs do Delivery Center.
- **`orders`**: informações sobre as vendas processadas através da plataforma do Delivery Center.
- **`payments`**: informações sobre os pagamentos realizados ao Delivery Center.
- **`stores`**: informações sobre os lojistas que utilizam a plataforma do Delivery Center para vender seus itens (*good* e/ou *food*) nos marketplaces.

## Dicionário de Dados

In [3]:
## Dicionário de Dados
dictionary = '../data/external/dicionario.csv'
dt_dict = pd.read_csv(dictionary)

# Exibe dicionário
dt_dict

Unnamed: 0                      variavel  \
0            0                    channel_id   
1            1                  channel_name   
2            2                  channel_type   
3            3                   delivery_id   
4            4             delivery_order_id   
5            5                     driver_id   
6            6      delivery_distance_meters   
7            7               delivery_status   
8            8                  driver_modal   
9            9                   driver_type   
10          10                        hub_id   
11          11                      hub_name   
12          12                      hub_city   
13          13                     hub_state   
14          14                  hub_latitude   
15          15                 hub_longitude   
16          16                      order_id   
17          17                      store_id   
18          18                    channel_id   
19          19              payment_order_id   
20          20                  order_status   
21          21                  order_amount   
22          22            order_delivery_fee   
23          23           order_delivery_cost   
24          24          order_moment_created   
25          25         order_moment_accepted   
26          26            order_moment_ready   
27          27        order_moment_collected   
28          28        order_moment_delivered   
29          29       order_metric_cycle_time   
30          30     order_metric_transit_time   
31          31  order_metric_production_time   
32          32                    payment_id   
33          33              payment_order_id   
34          34                payment_amount   
35          35                   payment_fee   
36          36                payment_method   
37          37                payment_status   
38          38                      store_id   
39          39                        hub_id   
40          40                    store_name   
41          41                 store_segment   
42          42              store_plan_price   
43          43                store_latitude   
44          44               store_longitude   

                                            descricao          tipo   subtipo  
0                     Identificador do canal de venda   qualitativa   nominal  
1             Nome do canal (ex: App, WhatsApp, etc.)   qualitativa   nominal  
2          Tipo do canal (Próprio, Marketplace, etc.)   qualitativa   nominal  
3                      Identificador único da entrega   qualitativa   nominal  
4                  ID do pedido relacionado à entrega   qualitativa   nominal  
5            ID do motorista responsável pela entrega   qualitativa   nominal  
6                Distância total da entrega em metros  quantitativa  contínua  
7   Status final da entrega (ex: entregue, cancela...   qualitativa   nominal  
8   Modal de transporte do motorista (carro, moto,...   qualitativa   nominal  
9   Tipo de vínculo com o motorista (CLT, parceiro...   qualitativa   nominal  
10                     Identificador do hub logístico   qualitativa   nominal  
11                              Nome do hub logístico   qualitativa   nominal  
12                  Cidade onde o hub está localizado   qualitativa   nominal  
13                  Estado onde o hub está localizado   qualitativa   nominal  
14                     Latitude da localização do hub  quantitativa  contínua  
15                    Longitude da localização do hub  quantitativa  contínua  
16                            Identificador do pedido   qualitativa   nominal  
17                 ID da loja responsável pelo pedido   qualitativa   nominal  
18              ID do canal usado para fazer o pedido   qualitativa   nominal  
19             ID da transação de pagamento associada   qualitativa   nominal  
20     Status do pedido (finalizado, cancelado, etc.)   qualitativa   nominal  
21                              Val

O dicionário de dados acima descreve as principais variáveis presentes no dataset, bem como seus respectivos tipos e significados:

- Variáveis **qualitativas nominais** predominam no dataset, como channel_name, delivery_status, driver_type, hub_state, entre outras. Essas variáveis representam identificadores, categorias ou status sem ordem implícita.

- Variáveis **quantitativas contínuas**, como delivery_distance_meters, order_amount, hub_latitude e hub_longitude, permitem análises numéricas e agregações estatísticas.

- Cada tabela pode ser relacionada a outra por meio de **identificadores únicos**, como order_id, delivery_id, driver_id, store_id e channel_id.

In [4]:
#verifica o formato do dataset e se existem dados duplicados (linhas idênticas)
for tabela, df in tabelas.items():
    linhas_duplicadas = df[df.duplicated()]
    print(f"Formato do Dataset {tabela}: {df.shape}\n", 
          f"Número de linhas duplicadas: {linhas_duplicadas.shape}\n")

Formato do Dataset deliveries: (378843, 5)
 Número de linhas duplicadas: (0, 5)

Formato do Dataset stores: (951, 7)
 Número de linhas duplicadas: (0, 7)

Formato do Dataset orders: (368999, 29)
 Número de linhas duplicadas: (0, 29)

Formato do Dataset drivers: (4824, 3)
 Número de linhas duplicadas: (0, 3)

Formato do Dataset hubs: (32, 6)
 Número de linhas duplicadas: (0, 6)

Formato do Dataset payments: (400834, 6)
 Número de linhas duplicadas: (0, 6)

Formato do Dataset channels: (40, 3)
 Número de linhas duplicadas: (0, 3)



É possível ver nesse breve resumo algumas informações importantes sobre o conjunto de dados que irão ajudar nas análises posteriores, como, por exemplo, o tipo e a qualidade de cada variável. Por hora, temos o seguinte:

- **`channels`** é uma **tabela de dimensão**, com **40 linhas e 3 colunas**, **sem dados duplicados e sem nulos**. Representa os diferentes canais de venda disponíveis.

- **`deliveries`** é uma **tabela de fatos**, com **378.843 linhas e 5 colunas**, **sem duplicados**, mas **com algumas variáveis contendo dados nulos**. Relaciona-se diretamente com a execução das entregas.

- **`drivers`** é uma **tabela de dimensão**, com **4.824 linhas e 3 colunas**, **sem dados nulos ou duplicados**. Contém informações sobre os motoristas envolvidos nas entregas.

- **`orders`** é uma **tabela de fatos (a principal do dataset)**, com **368.999 linhas e 29 colunas**, **sem duplicações**, mas **com diversas variáveis com dados ausentes — alguns em quantidade significativa**. Essa tabela centraliza a maioria dos relacionamentos entre as demais.

- **`payments`** é uma **tabela de fatos**, com **400.834 linhas e 6 colunas**, **sem duplicações**, e **apenas uma variável com baixo número de dados faltantes**. Contém as transações financeiras ligadas aos pedidos.

- **`hubs`** é uma **tabela de dimensão**, com **32 linhas e 6 colunas**, **sem nulos ou duplicações**. Armazena os dados dos hubs logísticos.

- **`stores`** é uma **tabela de dimensão**, com **951 linhas e 7 colunas**, **sem duplicados**, porém **com alguns dados ausentes em algumas colunas**.

Pode-se observar que os datasets **não apresentaram nenhuma linha duplicada**, o que é um indicativo positivo de consistência na coleta ou ingestão dos dados. Com isso, é finalizada a **leitura primária** do conjunto de dados apresentado.

## 2. Perguntas de partida e hipóteses

## 3. Insights